clients/js/src/signed_calls.ts

```diff
@@ -21,8 +18,68 @@ const DEFAULT_GAS_LIMIT = 30_000_000;
 const DEFAULT_VALUE = 0;
 const DEFAULT_DATA = '0x';
 const zeroAddress = () => `0x${'0'.repeat(40)}`;
+const MAX_SAFE_BIG_NUM = 0x1fffffffffffff - 1;
 
-export type Signer = Pick<EthersSigner, 'getTransactionCount' | 'getChainId'> &
+class SignedCallCacheManager {
+  public enabled = true;
+
+  // for each signer, we cache the signature of the hash of each SignableCall
+  private cachedSignatures = new Map<string, Map<string, Uint8Array>>();
+  // for each ChainId, we cache the base block number to make the same leash
+  private cachedBlockNumbers = new Map<number, BlockTag>();
+
+  public enable() {
+    this.enabled = true;
+  }
+
+  public disable() {
+    this.enabled = false;
+    cachedSignatures.clear();
+    cachedBlockNumbers.clear();
+  }
+
+  public setSignatureCache(
+    address: string,
+    hash: string,
+    signature: Uint8Array,
+  ) {
+    if (!this.enabled) return;
+    if (!this.cachedSignatures.has(address))
+      this.cachedSignatures.set(address, new Map<string, Uint8Array>());
+    this.cachedSignatures.get(address)!.set(hash, signature);
+  }
+
+  public getSignatureCache(
+    address: string,
+    hash: string,
+  ): Uint8Array | undefined {
+    return this.cachedSignatures.get(address)?.get(hash);
+  }
+
+  public setBlockNum(chainId: number, blockNum: BlockTag) {
+    if (!this.enabled) return;
+    this.cachedBlockNumbers.set(chainId, blockNum);
+  }
+
+  public getBlockNum(chainId: number): BlockTag | undefined {
+    return this.cachedBlockNumbers.get(chainId);
+  }
```

Comment:
```suggestion
```
Instead of doing it this way, please store the actual leash with the cached signature, then before sending, check if it's still valid, if so then send, otherwise regenerate the signature. If you do it this way, you end up being mostly unable to use the `blockRange` part of the leash.

Here's the algorithm I would suggest
1. digest the non-leash part of the call using either `_hashTypedData` or canonicalized JSON
2. check if the signature+leash is in the cache
3. check if the leash is still valid. you'll likely want a function called `leashHasSlack` to encapsulate this functionality
4. if the leash is still valid, attach the signature to the data pack and send. if the leash is invalid, invalidate the cache entry and recurse on `signCall`

This method will requiring checking the leash before sending, but makes it easier to handle short or custom `blockRange` and `baseBlock`s, which are both important for security.